include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -223,108 +218,155 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
 					std::array<uint64_t, 4> hash = {};
 					// If true, then `getIncludeStandard` was used to find, otherwise `getIncludeRelative`
 					bool standardInclude = false;
-					nbl::system::IFileBase::time_point_t lastWriteTime = {};
 				};
 
 				struct SMacroData {
 					std::string identifier;
 					std::string definition;
+					NLOHMANN_DEFINE_TYPE_INTRUSIVE(SMacroData, identifier, definition);
 				};
 
-				struct SPreprocessorData {
+				struct SCompilerArgs; // Forward declaration for SPreprocessorArgs's friend declaration
+
+				struct SPreprocessorArgs final {
 
-					inline bool operator==(const SPreprocessorData& other) const {
+					friend class SCompilerArgs;
+					friend class SEntry;
+
+					// Public default needed for json serialization
+					SPreprocessorArgs() {};
+					
+					inline bool operator==(const SPreprocessorArgs& other) const {
 						if (sourceIdentifier != other.sourceIdentifier) return false;
+
 						if (extraDefines.size() != other.extraDefines.size()) return false;
-						for (auto definesIt = extraDefines.begin(), otherDefinesIt = other.extraDefines.begin(); definesIt != extraDefines.end(); definesIt++, otherDefinesIt++) {
+						core::vector<SMacroData> sortedExtraDefines, otherSortedExtraDefines;
+
+						sortedExtraDefines.assign(extraDefines.begin(), extraDefines.end());
+						std::sort(sortedExtraDefines.begin(), sortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
+						
+						otherSortedExtraDefines.assign(other.extraDefines.begin(), other.extraDefines.end());
+						std::sort(otherSortedExtraDefines.begin(), otherSortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
```

Comment:
this is painful (memory allocation for a simple op)

You should instead sort the defines in the constructor  so this is amortized.

You will also be able to sort them in-place on the `extraDefines` vector instead of allocating new temporary memory.